filter.go

```diff
@@ -8,60 +8,162 @@ import (
 	manet "github.com/multiformats/go-multiaddr-net"
 )
 
+type filterEntry struct {
+	f      *net.IPNet
+	reject bool
+}
+
+// Filters is a structure representing a collection of allow/deny
+// net.IPNet filters, together with the RejectByDefault flag, which
+// represents the default filter policy.
+//
+// Note that the last policy added to the Filters is authoritative.
 type Filters struct {
+	RejectByDefault bool
+
 	mu      sync.RWMutex
-	filters map[string]*net.IPNet
+	filters []*filterEntry
 }
 
+// NewFilters constructs and returns a new set of net.IPNet filters.
+// By default, the new filter rejects no addresses.
 func NewFilters() *Filters {
 	return &Filters{
-		filters: make(map[string]*net.IPNet),
+		RejectByDefault: false,
+		filters:         make([]*filterEntry, 0),
 	}
 }
 
+func (fs *Filters) find(ff *net.IPNet) int {
+	ffs := ff.String()
+	for idx, ft := range fs.filters {
+		if ft.f.String() == ffs {
+			return idx
+		}
+	}
+
+	return -1
+}
+
+// AddDialFilter adds a reject rule to the given Filters.  Hosts
+// matching the given net.IPNet filter will be rejected, unless
+// another rule is added which states that they should be accepted.
+//
+// No effort is made to prevent duplication of filters, or to simplify
+// the filters list.
 func (fs *Filters) AddDialFilter(f *net.IPNet) {
```

Comment:
Let's mark this as DEPRECATED, move the logic to `AddDenyFilter` and make it call that.